### ISSUE TITLE
Date pipe 'HH:mm' produces memory leak

### ISSUE THREAD

**Issue by OddDev** [angular/angular#18579]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Using the date pipe like this {{time | date:'HH:mm' }} and updating "time", the pipe produces a memory leak. In our case we updated time every second. After about 15 hours we killed a reaspberry PI with 1 GB RAM (aprox. 15 * 3600 seconds consumed about 750 MB). 

## Expected behavior
To let the garbage collector clean up resources to free RAM.

## Minimal reproduction of the problem with instructions

Create an observable with an interval of 1000 ms. 
Subscribe to this observable to update a variable (class-field etc.) every second. 
This variable has to be piped through date. Only tested with the arguments 'HH:mm' but it's likely to fail with other configs as well. 
Watch the RAM consumption of the browser go up. 

Edit: Stackblitz-Link from the comment-section: https://stackblitz.com/edit/angular-odb6rt

## What is the motivation / use case for changing the behavior?

The date pipe could kill the client if it runs on a device with small resources or if the app runs over a long period of time. 

## Environment

<pre><code>
Angular version: 4.0.0

Browser:
- [x] Chrome (desktop) version 60
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [x] Firefox version 54
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 6.10.1 x64  
- Platform:   Linux and Windows

Others:

</code></pre>

**Comment by Toxicable** [comment]
I attempted to reproduce your issue but was unsuccessful
I used this https://stackblitz.com/edit/angular-template-rpev6k?file=app%2Fapp.module.ts
and ran it for 23 minutes and I did not see any memory leaks, note im doing a 20ms interval instead of your 1000ms interval.
![image](https://user-images.githubusercontent.com/13490925/29062925-c633c87e-7c77-11e7-9cbe-9e11c13f9053.png)
And here's the Heap Allocation timeline.
[Heap-20170808T202221.zip]()
If you are able to reproduce this issue please link a plunkr or stackblitz with it the reproduction, otherwise this is unactionable and more likely to be an issue with your application

**Comment by OddDev** [comment]
This eats my RAM successfully: https://stackblitz.com/edit/angular-odb6rt :)

**Comment by OddDev** [comment]
We've implemented our own pipe like this to work around this issue: 

```
import {Pipe, PipeTransform} from '@angular/core';

const HOUR_MINUTE_CONFIG = '2-digit';

const DATE_FORMAT_OPTIONS = {
  hour: HOUR_MINUTE_CONFIG,
  minute: HOUR_MINUTE_CONFIG
};


@Pipe({
  name: 'commonClock'
})
export class CommonClockPipe implements PipeTransform {

  transform(value: Date): any {
    return value.toLocaleTimeString([], DATE_FORMAT_OPTIONS);
  }

}
```

Try this code and you'll recognize that the RAM doesn't get eaten up anymore.

**Comment by Toxicable** [comment]
I tried to reproduce your issue again but was unsuccessful
I adjusted the clock to 20ms to expedite the process and ran it for 28 minutes.
![image](https://user-images.githubusercontent.com/13490925/29065953-e4a0756e-7c81-11e7-97e3-dd6a62fae0cb.png)
As you can see no leaks.
That's using this plunkr here https://stackblitz.com/edit/angular-mrulwv?file=app/app.component.ts
Could you confirm that your reproduction does leak memory through a Heap allocation file or similar

**Comment by OddDev** [comment]
You need to check the process of the browser. My colleague just pointed out that he had also not been able to capture the issue with analysis tools. You need to monitor the chrome processes.

**Comment by OddDev** [comment]
Or simply put, just monitor the RAM consumption of the client running the app.

**Comment by borislemke** [comment]
Like @Toxicable , I couldn't reproduce the issue either with your `stackblitz` snippets. I increased the interval to 20ms as well. I ran it on a mac with 8GB of RAM for 30 mins. My RAM consumption is steady.

**Comment by ExTheSea** [comment]
That's weird as we could reliably reproduce it on both our Windows 7 work machines using the latest Google Chrome and Firefox and also the Chromium browser running on a Raspberry PI 3 setup with Raspbian. 

In all cases we noticed a steady increase in the memory usage of the Browser we were testing once we opened a basic Angular application using the date pipe. We were checking this by looking at the processes in the Windows Task manager or respectively top on the Raspberry.   
On the Raspberry it would even go as far as too crash the entire system by taking up to much RAM.

We are certain it has to do with the date picker as the exact same application without the date pipe and either no pipe at all or the simple alternative pipe @OddDev mentioned, would not further increase the memory usage after the website load. Without the date pipe the Raspberry is currently running with a steadily low memory usage for several weeks now, the only difference being the pipe.
